{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Are there sequences near TSS that are associated with up or down regulation by ...  \n",
    "    Call transcripts from merged RNAseq datasets, find TSS region,\n",
    "    Merge RNAseq data to obtain more sequencing depth \n",
    "    Merge .bam files from BAO and BAN experimental set to one bam file\n",
    "    split merged bam file into + strand and - strand files (+ flag 83,163) (-flag 99,147)\n",
    "    index each file\n",
    "    Run stringtie on this indexed + and - bam file to estimate transcripts\n",
    "\n",
    "\n",
    "    Filtering out stringtie transcripts that are more reliable and editing them\n",
    "    filter\n",
    "    Select stringtie transcripts that start between two genes on one of the\n",
    "    DNA strands\n",
    "    Filter out transcripts that\n",
    "    Have low read Density\n",
    "    have no sudden increase in reads over a window of nts\n",
    "    Edit \n",
    "    If stringtie transcript ends in middle of gene, extend transcript to end of that gene\n",
    "    Create GFF from Filtered/edited stringtie transcripts and visualize with genome browser\n",
    "\n",
    "    match known transcripts  with their known TSSs to determine if our predictions \n",
    "    are predicting those correctly\n",
    "\n",
    "    generate file that has:\n",
    "    all predicted TSS sequences from -100 to +50 (wrt TSS)\n",
    "    genes that are included in TU for each TSS\n",
    "    information about whether those genes are upregulated, downregulated, not regulated \n",
    "    TSS1\t….NNNNNN….\trpoD\tnot regulated\n",
    "    TSS1\t…NNNNNN….\tlecA\tnot regulated\n",
    "    TSS2\t…NNNNNN…\t\tdnaA\tupregulated\n",
    "    \n",
    "    determine if there are motifs present in upregulated, downregulated or not regulated \n",
    "    seq using MEME\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from plotly import offline as pyo\n",
    "from plotly import graph_objects as go\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import ndimage as ndi\n",
    "from jw_utils import parse_gff as pgf\n",
    "from jw_utils import parse_fasta as pf\n",
    "from jw_utils import file_utils as fu\n",
    "from Bio import SeqIO\n",
    "import bisect\n",
    "import pysam\n",
    "from transcript_calling import tc_functions as tf\n",
    "from jw_utils import genome_utils as gu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_gff = '../references/FERM_BP3421.gff'\n",
    "path_to_strGTF_neg = './merge_all_negStrand.gtf'\n",
    "path_to_annot_file = '../data/references/Reference_FERM_BP3421.gbk'\n",
    "path_to_fa_genomes = '../data/references/concat_references.fa'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split merged bam file into plus and minus strand templates \n",
    "    stringtie seems to perform better for me when I give it only one strand\n",
    "    - rf sequencing  \n",
    "    - plus strand flags: 83, 163\n",
    "    - minus strand flags: 99, 147\n",
    "    samtools view -f 83 merge_all.bam -o flag83.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: samtools: command not found\n",
      "/bin/bash: samtools: command not found\n",
      "/bin/bash: samtools: command not found\n",
      "rm: ./merged_bam_files/flag83.bam: No such file or directory\n",
      "rm: ./merged_bam_files/flag163.bam: No such file or directory\n",
      "/bin/bash: samtools: command not found\n",
      "/bin/bash: samtools: command not found\n",
      "/bin/bash: samtools: command not found\n",
      "/bin/bash: samtools: command not found\n",
      "rm: ./merged_bam_files/flag99.bam: No such file or directory\n",
      "rm: ./merged_bam_files/flag147.bam: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "#positive strand genes\n",
    "!samtools view -f 83 ./merged_bam_files/merge_all.bam -o ./merged_bam_files/flag83.bam\n",
    "#!samtools view -f 163 merge_all.bam -o flag163.bam\n",
    "!samtools merge ./merged_bam_files/flag83.bam ./merged_bam_files/flag163.bam -o ./merged_bam_files/merge_all_posStrand.bam\n",
    "!samtools index ./merged_bam_files/merge_all_posStrand.bam\n",
    "!rm ./merged_bam_files/flag83.bam\n",
    "!rm ./merged_bam_files/flag163.bam\n",
    "\n",
    "#neg strand genes\n",
    "!samtools view -f 99 ./merged_bam_files/merge_all.bam -o ./merged_bam_files/flag99.bam\n",
    "!samtools view -f 147 ./merged_bam_files/merge_all.bam -o ./merged_bam_files/flag147.bam\n",
    "!samtools merge ./merged_bam_files/flag99.bam ./merged_bam_files/flag147.bam -o ./merged_bam_files/merge_all_negStrand.bam\n",
    "!samtools index ./merged_bam_files/merge_all_negStrand.bam\n",
    "!rm ./merged_bam_files/flag99.bam\n",
    "!rm ./merged_bam_files/flag147.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call initial transcripts using stringtie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "!stringtie -g 0 -j 10 ./merged_bam_files/merge_all_posStrand.bam --rf -o merge_all_posStrand.gtf\n",
    "!stringtie -g 0 -j 10 ./merged_bam_files/merge_all_negStrand.bam --rf -o merge_all_negStrand.gtf\n",
    "!stringtie -g 0 -j 10 ./merged_bam_files/merge_all.bam --rf -o merge_all.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Merge pos and neg strand annotation files produced by stringtie back into one gff file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_nested_obj_dict = tf.merge_pos_neg_sttie_objs('merge_all_posStrand.gtf','merge_all_negStrand.gtf',path_to_gff)\n",
    "tf.write_gff_from_seqobj(merged_nested_obj_dict,'merged_all.gff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make wig files with pilups for each strand for each chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"./merged_bam_files/merge_all_posStrand.bam\" : No such file or directory\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] could not open alignment file `./merged_bam_files/merge_all_posStrand.bam`: No such file or directory",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/hb/rvrzm3nx3sxcpzvvy7xxwtsw0000gn/T/ipykernel_92983/3614808096.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcontig_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpgf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_contig_names\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_to_gff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake_wig_pilups\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./merged_bam_files/merge_all_posStrand.bam'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontig_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake_wig_pilups\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./merged_bam_files/merge_all_negStrand.bam'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontig_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Trestle_projects/Eustaquio_lab/Eustaquio_epigenetics/project_1/transcript_calling/tc_functions.py\u001b[0m in \u001b[0;36mmake_wig_pilups\u001b[0;34m(bam_filepath, contig_names)\u001b[0m\n\u001b[1;32m    167\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmake_wig_pilups\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_filepath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontig_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 169\u001b[0;31m     \u001b[0msam_align_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpysam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAlignmentFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_filepath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    170\u001b[0m     \u001b[0mfilename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbam_filepath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.bam'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mcontig\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcontig_names\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpysam/libcalignmentfile.pyx\u001b[0m in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpysam/libcalignmentfile.pyx\u001b[0m in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile._open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] could not open alignment file `./merged_bam_files/merge_all_posStrand.bam`: No such file or directory"
     ]
    }
   ],
   "source": [
    "contig_names = pgf.get_contig_names(path_to_gff)\n",
    "tf.make_wig_pilups('./merged_bam_files/merge_all_posStrand.bam', contig_names)\n",
    "tf.make_wig_pilups('./merged_bam_files/merge_all_negStrand.bam', contig_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering out stringtie transcripts that are more reliable, and editing them    \n",
    "\n",
    "    filter for  \n",
    "    -Select stringtie transcripts that start between two genes on one of the    \n",
    "     DNA strands  \n",
    "    -Filter out transcripts that  \n",
    "        -Have low read Density  \n",
    "        -have no sudden increase in reads over a window of nts  \n",
    "    Edit   \n",
    "        -If stringtie transcript ends in middle of gene, extend transcript to  \n",
    "         end of that gene  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Edit  and filter transcripts  \n",
    "tf.filter_transcripts() \n",
    "1) extends each transcript to the stop codon of the nearest gene on the same strand\n",
    "2) filters out transcripts that start in the middle of a gene (on the same strand)\n",
    "3) filters out genes that have below the input threshold reads/nt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_transcript_objs = tf.filter_transcripts('./merged_all.gff',path_to_gff,20,annot_type='gff' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf.write_gff_from_seqobj(filtered_transcript_objs, 'merged_all_filtered_20.gff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get sequence around a given coordinate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div>\n",
    "<img src=\"./notebook_images/get_minusStrand_seq_region2.jpg\" alt=\"get minus strand seq. region\" width=\"400\">\n",
    "<div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Generate file that has:  \n",
    "- all predicted TSS sequences from -100 to +50 (wrt TSS)\n",
    "- genes that are included in TU for each TSS\n",
    "- information about whether those genes are upregulated, downregulated, not regulated \n",
    "        TSS1    ….NNNNNN….    rpoD    not regulated\n",
    "        TSS1    …NNNNNN….    lecA    not regulated\n",
    "        TSS2    …NNNNNN…        dnaA    upregulated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not return all of upstream sequence requested because reached beginning of contig.\n"
     ]
    }
   ],
   "source": [
    "def make_seqs_around_start_df(upstream, downstream, fasta_genome, feature_obj_dict):\n",
    "    \"\"\"\n",
    "    \n",
    "    parameters:\n",
    "    upstream (int): # nt to return upstream of feature_beginning\n",
    "    downstream (int): # nt to return downstream of feature_beginning\n",
    "    fasta_genome (str): path_to_fasta_genome\n",
    "    feature_obj_dict (dict): {contig:feature_ID:feature_annot_object}\n",
    "    \"\"\"\n",
    "    info_for_file = {}\n",
    "#     upstream = 100\n",
    "#     downstream = 50\n",
    "    fasta_genome_dict = pf.get_seq_dict(fasta_genome)\n",
    "    seqs = {}\n",
    "    contigs = {}\n",
    "    for contig in filtered_transcript_objs:\n",
    "        for seq_obj_id in filtered_transcript_objs[contig]:\n",
    "            seq_obj = filtered_transcript_objs[contig][seq_obj_id]\n",
    "            strand = seq_obj.strand\n",
    "            if strand == '+':\n",
    "                tss = seq_obj.start\n",
    "            elif strand == '-':\n",
    "                tss = seq_obj.end\n",
    "            contigs[seq_obj_id] = contig\n",
    "            seqs[seq_obj_id] = tf.get_seq_around_coordinate(fasta_genome_dict,\n",
    "                                            contig,strand, tss, upstream, downstream)\n",
    "            \n",
    "    df1 = pd.DataFrame(seqs.values(), seqs.keys())\n",
    "    df2 = pd.DataFrame(contigs.values(), contigs.keys()) \n",
    "    df_seqs= pd.merge(df1,df2,how='inner', left_index=True, right_index=True)\n",
    "    df_seqs.columns =[\"up_down_seq 3'-5'\", 'chromosome']\n",
    "    df_seqs.index.name = 'transcript_ID'\n",
    "    return df_seqs\n",
    "            \n",
    "fasta_concat_genomepath = '../references/concat_references.fa'            \n",
    "df_seqs = make_seqs_around_start_df(100, 50, fasta_concat_genomepath, filtered_transcript_objs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>up_down_seq 3'-5'</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>STRG.2363</th>\n",
       "      <td>aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...</td>\n",
       "      <td>BF000000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.2366</th>\n",
       "      <td>aatactcccccaatttatcttgacaatcgttaacaaatgcccgccg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.2370</th>\n",
       "      <td>tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.9</th>\n",
       "      <td>caacggatgccgtccgccgatatccgatgcccggacctggcgtcga...</td>\n",
       "      <td>BF000000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.12</th>\n",
       "      <td>ccgaacggcgacgcgtacaccaggtagtcgaccgccatgtcgcgcg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.3293</th>\n",
       "      <td>gcgcgcggcgcgggccgggtgcgcgggatctgagcccgatcagaaa...</td>\n",
       "      <td>BF000000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.3297</th>\n",
       "      <td>tggtcggcggcgatcttcgcgccccagttcgcccgcgaggcggggc...</td>\n",
       "      <td>BF000000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.3298</th>\n",
       "      <td>cgaagatggtcagccggttcagattgacgtcgcgcaggtccgccat...</td>\n",
       "      <td>BF000000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.3300</th>\n",
       "      <td>ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...</td>\n",
       "      <td>BF000000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRG.3309</th>\n",
       "      <td>gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...</td>\n",
       "      <td>BF000000.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1133 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               up_down_seq 3'-5'  chromosome\n",
       "transcript_ID                                                               \n",
       "STRG.2363      aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...  BF000000.3\n",
       "STRG.2366      aatactcccccaatttatcttgacaatcgttaacaaatgcccgccg...  BF000000.3\n",
       "STRG.2370      tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...  BF000000.3\n",
       "STRG.9         caacggatgccgtccgccgatatccgatgcccggacctggcgtcga...  BF000000.3\n",
       "STRG.12        ccgaacggcgacgcgtacaccaggtagtcgaccgccatgtcgcgcg...  BF000000.3\n",
       "...                                                          ...         ...\n",
       "STRG.3293      gcgcgcggcgcgggccgggtgcgcgggatctgagcccgatcagaaa...  BF000000.2\n",
       "STRG.3297      tggtcggcggcgatcttcgcgccccagttcgcccgcgaggcggggc...  BF000000.2\n",
       "STRG.3298      cgaagatggtcagccggttcagattgacgtcgcgcaggtccgccat...  BF000000.2\n",
       "STRG.3300      ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...  BF000000.2\n",
       "STRG.3309      gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...  BF000000.2\n",
       "\n",
       "[1133 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get genes within the transcript  \n",
    "1) if the start codon of a gene (on the same strand as the transcript) is within the \n",
    "    start:end coordinates of the transcript, then count it as in the gene\n",
    "    - need the gene annotation  and the transcript annotations\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>up_down_seq 3'-5'</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genes</th>\n",
       "      <th>#_genes_in_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STRG.2363</td>\n",
       "      <td>aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...</td>\n",
       "      <td>BF000000.3</td>\n",
       "      <td>gene-tmp_006144</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STRG.2363</td>\n",
       "      <td>aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...</td>\n",
       "      <td>BF000000.3</td>\n",
       "      <td>gene-tmp_006145</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STRG.2366</td>\n",
       "      <td>aatactcccccaatttatcttgacaatcgttaacaaatgcccgccg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "      <td>gene-tmp_006149</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STRG.2370</td>\n",
       "      <td>tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "      <td>gene-tmp_006157</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STRG.2370</td>\n",
       "      <td>tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...</td>\n",
       "      <td>BF000000.3</td>\n",
       "      <td>gene-tmp_006158</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1777</th>\n",
       "      <td>STRG.3300</td>\n",
       "      <td>ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...</td>\n",
       "      <td>BF000000.2</td>\n",
       "      <td>gene-tmp_006046</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1778</th>\n",
       "      <td>STRG.3300</td>\n",
       "      <td>ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...</td>\n",
       "      <td>BF000000.2</td>\n",
       "      <td>gene-tmp_006047</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1779</th>\n",
       "      <td>STRG.3309</td>\n",
       "      <td>gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...</td>\n",
       "      <td>BF000000.2</td>\n",
       "      <td>gene-tmp_006102</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1780</th>\n",
       "      <td>STRG.3309</td>\n",
       "      <td>gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...</td>\n",
       "      <td>BF000000.2</td>\n",
       "      <td>gene-tmp_006103</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1781</th>\n",
       "      <td>STRG.3309</td>\n",
       "      <td>gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...</td>\n",
       "      <td>BF000000.2</td>\n",
       "      <td>gene-tmp_006104</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1782 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_ID                                  up_down_seq 3'-5'  \\\n",
       "0        STRG.2363  aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...   \n",
       "1        STRG.2363  aacatcgaattaattctccgcccaacaaatatctaaaatccaatgt...   \n",
       "2        STRG.2366  aatactcccccaatttatcttgacaatcgttaacaaatgcccgccg...   \n",
       "3        STRG.2370  tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...   \n",
       "4        STRG.2370  tgaaggtcggcgtccgcgacggacaaatcgacgcagcgcaccggcg...   \n",
       "...            ...                                                ...   \n",
       "1777     STRG.3300  ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...   \n",
       "1778     STRG.3300  ccctgggagtgaagatttcgagcccgaggctgatataccgatggcg...   \n",
       "1779     STRG.3309  gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...   \n",
       "1780     STRG.3309  gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...   \n",
       "1781     STRG.3309  gaggcggccgccgaagcgcgaaaacacccccgcgagcatgcgcgga...   \n",
       "\n",
       "      chromosome            genes  #_genes_in_transcript  \n",
       "0     BF000000.3  gene-tmp_006144                      2  \n",
       "1     BF000000.3  gene-tmp_006145                      2  \n",
       "2     BF000000.3  gene-tmp_006149                      1  \n",
       "3     BF000000.3  gene-tmp_006157                      2  \n",
       "4     BF000000.3  gene-tmp_006158                      2  \n",
       "...          ...              ...                    ...  \n",
       "1777  BF000000.2  gene-tmp_006046                      4  \n",
       "1778  BF000000.2  gene-tmp_006047                      4  \n",
       "1779  BF000000.2  gene-tmp_006102                      3  \n",
       "1780  BF000000.2  gene-tmp_006103                      3  \n",
       "1781  BF000000.2  gene-tmp_006104                      3  \n",
       "\n",
       "[1782 rows x 5 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_genes_in_transcripts_df(path_trans_annots, path_to_gff):\n",
    "    \n",
    "    genes_in_transcript=tf.find_genes_in_transcript(path_trans_annots, path_to_gff)\n",
    "    transcpt_list = []\n",
    "    gene_list = []\n",
    "    num_genes = []\n",
    "    df_git =pd.DataFrame(genes_in_transcript.values(), genes_in_transcript.keys())\n",
    "    for tnscpt, genes in genes_in_transcript.items():\n",
    "        if len(genes)>0:\n",
    "            for gene in genes:\n",
    "                transcpt_list.append(tnscpt)\n",
    "                gene_list.append(gene)\n",
    "                num_genes.append(len(genes))\n",
    "        else:\n",
    "            transcpt_list.append(tnscpt)\n",
    "            gene_list.append(None)\n",
    "            num_genes.append(0)\n",
    "\n",
    "    df_tran_genes = pd.DataFrame()\n",
    "    df_tran_genes['transcript_ID'] = transcpt_list\n",
    "    df_tran_genes['genes'] = gene_list\n",
    "    df_tran_genes['#_genes_in_transcript'] = num_genes\n",
    "    return df_tran_genes\n",
    "\n",
    "path_trans_annots = './merged_all_filtered_20.gff'\n",
    "df_tran_genes = get_genes_in_transcripts_df(path_trans_annots, path_to_gff)\n",
    "df_merge = pd.merge(df_seqs,df_tran_genes, how='outer', on='transcript_ID')\n",
    "df_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add gene expression data (up and down-regulated genes) to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def add_gene_expression_data(data_path, up_thresh, down_thresh, pval_thresh):\n",
    "    df_expression = pd.read_csv(data_path, sep='\\t')\n",
    "    pval_filt = df_expression['PValue']<pval_thresh\n",
    "    df_expression = df_expression.loc[pval_filt,:]\n",
    "    filt_up = df_expression['logFC']>up_thresh\n",
    "    filt_down = df_expression['logFC']<down_thresh\n",
    "    up_down = []\n",
    "    for up,down in zip(filt_up,filt_down):\n",
    "        if up:\n",
    "            up_down.append('up')\n",
    "        elif down:\n",
    "            up_down.append('down')\n",
    "        else:\n",
    "            up_down.append('not significant')\n",
    "    df_expression['regulation'] = up_down\n",
    "    return df_expression\n",
    "\n",
    "data_path = './differentially_regulated_genes_BANvBAO.tsv'\n",
    "df_expression_full = add_gene_expression_data(data_path,1.0, -1.0,5)\n",
    "df_expression = add_gene_expression_data(data_path,1.5, -1.5,0.00005)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Locustag</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Description</th>\n",
       "      <th>FeatureType</th>\n",
       "      <th>logFC</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>BAO-1</th>\n",
       "      <th>BAO-2</th>\n",
       "      <th>BAO-3</th>\n",
       "      <th>BAN-1</th>\n",
       "      <th>BAN-2</th>\n",
       "      <th>BAN-3</th>\n",
       "      <th>regulation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tmp_000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>helix-turn-helix transcriptional regulator</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1.026274</td>\n",
       "      <td>1.177951e-02</td>\n",
       "      <td>3.269609e-02</td>\n",
       "      <td>10.997404</td>\n",
       "      <td>22.632058</td>\n",
       "      <td>10.281642</td>\n",
       "      <td>35.975102</td>\n",
       "      <td>20.201871</td>\n",
       "      <td>32.953047</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tmp_000007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transglycosylase domain-containing protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-1.101383</td>\n",
       "      <td>6.022556e-03</td>\n",
       "      <td>1.917286e-02</td>\n",
       "      <td>57.574643</td>\n",
       "      <td>45.264115</td>\n",
       "      <td>46.909989</td>\n",
       "      <td>9.734439</td>\n",
       "      <td>32.498662</td>\n",
       "      <td>27.351029</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tmp_000010</td>\n",
       "      <td>groL</td>\n",
       "      <td>chaperonin GroEL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-1.017040</td>\n",
       "      <td>1.352552e-05</td>\n",
       "      <td>1.437962e-04</td>\n",
       "      <td>14840.996203</td>\n",
       "      <td>17186.647460</td>\n",
       "      <td>20057.554761</td>\n",
       "      <td>9452.987059</td>\n",
       "      <td>7495.772396</td>\n",
       "      <td>8787.918446</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tmp_000012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rubredoxin</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-1.101685</td>\n",
       "      <td>1.049378e-04</td>\n",
       "      <td>7.422253e-04</td>\n",
       "      <td>973.270222</td>\n",
       "      <td>1072.279461</td>\n",
       "      <td>1018.203810</td>\n",
       "      <td>600.149346</td>\n",
       "      <td>498.459203</td>\n",
       "      <td>329.530465</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tmp_000040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oligosaccharide flippase family protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-1.009898</td>\n",
       "      <td>9.346225e-03</td>\n",
       "      <td>2.723772e-02</td>\n",
       "      <td>23.288619</td>\n",
       "      <td>21.260418</td>\n",
       "      <td>14.137257</td>\n",
       "      <td>10.157676</td>\n",
       "      <td>7.905080</td>\n",
       "      <td>10.874505</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1522</th>\n",
       "      <td>tmp_006781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside O-phosphotransferase APH(3')-IIa</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-4.156416</td>\n",
       "      <td>4.462669e-14</td>\n",
       "      <td>7.638601e-11</td>\n",
       "      <td>54483.078317</td>\n",
       "      <td>50513.038257</td>\n",
       "      <td>52188.005754</td>\n",
       "      <td>2842.456283</td>\n",
       "      <td>3100.547991</td>\n",
       "      <td>2871.528474</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523</th>\n",
       "      <td>tmp_006782</td>\n",
       "      <td>traJ</td>\n",
       "      <td>conjugal transfer transcriptional regulator TraJ</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-4.545283</td>\n",
       "      <td>1.394663e-13</td>\n",
       "      <td>1.790398e-10</td>\n",
       "      <td>11431.801077</td>\n",
       "      <td>11664.768289</td>\n",
       "      <td>11780.512051</td>\n",
       "      <td>563.751008</td>\n",
       "      <td>487.479925</td>\n",
       "      <td>442.559415</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>tmp_006783</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TraK family protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-4.341258</td>\n",
       "      <td>3.084284e-13</td>\n",
       "      <td>3.167560e-10</td>\n",
       "      <td>12245.932399</td>\n",
       "      <td>10315.074666</td>\n",
       "      <td>10939.987859</td>\n",
       "      <td>496.879642</td>\n",
       "      <td>615.717887</td>\n",
       "      <td>540.100432</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>tmp_006784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-4.586525</td>\n",
       "      <td>2.565692e-05</td>\n",
       "      <td>2.439783e-04</td>\n",
       "      <td>9.703591</td>\n",
       "      <td>12.001849</td>\n",
       "      <td>8.996436</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.988591</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>tmp_006785</td>\n",
       "      <td>NaN</td>\n",
       "      <td>replication initiation protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td>-4.301657</td>\n",
       "      <td>1.929591e-12</td>\n",
       "      <td>1.415493e-09</td>\n",
       "      <td>1498.881425</td>\n",
       "      <td>1405.245038</td>\n",
       "      <td>1593.654433</td>\n",
       "      <td>79.568461</td>\n",
       "      <td>84.760023</td>\n",
       "      <td>63.928910</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1527 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Locustag  Gene                                       Description  \\\n",
       "0     tmp_000001   NaN        helix-turn-helix transcriptional regulator   \n",
       "1     tmp_000007   NaN        transglycosylase domain-containing protein   \n",
       "2     tmp_000010  groL                                  chaperonin GroEL   \n",
       "3     tmp_000012   NaN                                        rubredoxin   \n",
       "4     tmp_000040   NaN           oligosaccharide flippase family protein   \n",
       "...          ...   ...                                               ...   \n",
       "1522  tmp_006781   NaN   aminoglycoside O-phosphotransferase APH(3')-IIa   \n",
       "1523  tmp_006782  traJ  conjugal transfer transcriptional regulator TraJ   \n",
       "1524  tmp_006783   NaN                               TraK family protein   \n",
       "1525  tmp_006784   NaN                              hypothetical protein   \n",
       "1526  tmp_006785   NaN                    replication initiation protein   \n",
       "\n",
       "     FeatureType     logFC        PValue           FDR         BAO-1  \\\n",
       "0            CDS  1.026274  1.177951e-02  3.269609e-02     10.997404   \n",
       "1            CDS -1.101383  6.022556e-03  1.917286e-02     57.574643   \n",
       "2            CDS -1.017040  1.352552e-05  1.437962e-04  14840.996203   \n",
       "3            CDS -1.101685  1.049378e-04  7.422253e-04    973.270222   \n",
       "4            CDS -1.009898  9.346225e-03  2.723772e-02     23.288619   \n",
       "...          ...       ...           ...           ...           ...   \n",
       "1522         CDS -4.156416  4.462669e-14  7.638601e-11  54483.078317   \n",
       "1523         CDS -4.545283  1.394663e-13  1.790398e-10  11431.801077   \n",
       "1524         CDS -4.341258  3.084284e-13  3.167560e-10  12245.932399   \n",
       "1525         CDS -4.586525  2.565692e-05  2.439783e-04      9.703591   \n",
       "1526         CDS -4.301657  1.929591e-12  1.415493e-09   1498.881425   \n",
       "\n",
       "             BAO-2         BAO-3        BAN-1        BAN-2        BAN-3  \\\n",
       "0        22.632058     10.281642    35.975102    20.201871    32.953047   \n",
       "1        45.264115     46.909989     9.734439    32.498662    27.351029   \n",
       "2     17186.647460  20057.554761  9452.987059  7495.772396  8787.918446   \n",
       "3      1072.279461   1018.203810   600.149346   498.459203   329.530465   \n",
       "4        21.260418     14.137257    10.157676     7.905080    10.874505   \n",
       "...            ...           ...          ...          ...          ...   \n",
       "1522  50513.038257  52188.005754  2842.456283  3100.547991  2871.528474   \n",
       "1523  11664.768289  11780.512051   563.751008   487.479925   442.559415   \n",
       "1524  10315.074666  10939.987859   496.879642   615.717887   540.100432   \n",
       "1525     12.001849      8.996436     0.000000     0.000000     0.988591   \n",
       "1526   1405.245038   1593.654433    79.568461    84.760023    63.928910   \n",
       "\n",
       "     regulation  \n",
       "0            up  \n",
       "1          down  \n",
       "2          down  \n",
       "3          down  \n",
       "4          down  \n",
       "...         ...  \n",
       "1522       down  \n",
       "1523       down  \n",
       "1524       down  \n",
       "1525       down  \n",
       "1526       down  \n",
       "\n",
       "[1527 rows x 14 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expression_subset = df_expression[['Locustag','Description','regulation','PValue','logFC']]\n",
    "df_expression_subset.columns = ['genes','Description','regulation','PValue','logFC']\n",
    "df_merge['genes'] = [gene.replace('gene-', '') if gene else None for gene in df_merge['genes']]\n",
    "df3 = pd.merge(df_merge,df_expression_subset, how='inner', on='genes').set_index('transcript_ID')\n",
    "df3.to_csv('full.csv')\n",
    "df_expression_full\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Volcano plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2        2.033241\n",
       "5        2.664175\n",
       "9       12.077312\n",
       "10      10.234666\n",
       "11       0.200274\n",
       "          ...    \n",
       "1522    18.503156\n",
       "1523    24.151807\n",
       "1524    21.638660\n",
       "1525    14.008044\n",
       "1526    22.592076\n",
       "Length: 617, dtype: float64"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc_select"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Graph read density on filtered transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'coverage_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/hb/rvrzm3nx3sxcpzvvy7xxwtsw0000gn/T/ipykernel_96624/2280616836.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mfilt_val\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m20\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mcoverage_fil_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcoverage_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcoverage_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'coverage'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0mfilt_val\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'coverage'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mbins\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0;36m50\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mrnge\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m75\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'coverage_df' is not defined"
     ]
    }
   ],
   "source": [
    "filt_val = 20\n",
    "coverage_fil_df = coverage_df.loc[coverage_df['coverage']>filt_val,:].sort_values('coverage')\n",
    "\n",
    "bins  = 50\n",
    "rnge = (0,75)\n",
    "import plotly.graph_objects as go\n",
    "import plotly.offline as pyo\n",
    "hist = np.histogram(coverage_fil_df['coverage'],bins=bins, range=rnge)\n",
    "trace1 = go.Bar(x=hist[1], y=hist[0], name='coverage')\n",
    "layout = go.Layout({'title':'histogram of ave. reads/nt'})\n",
    "fig = go.Figure(data=[trace1], layout=layout )\n",
    "fig.update_xaxes({'title':\n",
    "                      {'text':'ave. reads/nt',\n",
    "                      'font':{'size':20}}\n",
    "                 })\n",
    "fig.update_yaxes({'title':\n",
    "                      {'text':'number of transripts',\n",
    "                      'font':{'size':20}}\n",
    "                 })\n",
    "fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "vscode": {
   "interpreter": {
    "hash": "6691013dd57765de18105885afeee5e2162f23612599fe49b9f063fc824dadcf"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
